format/pyrdp/pdu/clipboard_data.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2022-2023 GoSecure Inc.
+// Copyright (c) 2024 Flare Systems
+// Licensed under the MIT License
+package pdu
+
+import (
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+const (
+	// Message types.
+	CB_TYPE_MONITOR_READY         = 0x0001
+	CB_TYPE_FORMAT_LIST           = 0x0002
+	CB_TYPE_FORMAT_LIST_RESPONSE  = 0x0003
+	CB_TYPE_FORMAT_DATA_REQUEST   = 0x0004
+	CB_TYPE_FORMAT_DATA_RESPONSE  = 0x0005
+	CB_TYPE_TEMP_DIRECTORY        = 0x0006
+	CB_TYPE_CLIP_CAPS             = 0x0007
+	CB_TYPE_FILECONTENTS_REQUEST  = 0x0008
+	CB_TYPE_FILECONTENTS_RESPONSE = 0x0009
+	CB_TYPE_LOCK_CLIPDATA         = 0x000a
+	CB_TYPE_UNLOCK_CLIPDATA       = 0x000b
+)
+
+var cbTypesMap = scalar.UintMapSymStr{
+	CB_TYPE_MONITOR_READY:         "monitor_ready",
+	CB_TYPE_FORMAT_LIST:           "format_list",
+	CB_TYPE_FORMAT_LIST_RESPONSE:  "format_list_response",
+	CB_TYPE_FORMAT_DATA_REQUEST:   "format_data_request",
+	CB_TYPE_FORMAT_DATA_RESPONSE:  "format_data_response",
+	CB_TYPE_TEMP_DIRECTORY:        "temp_directory",
+	CB_TYPE_CLIP_CAPS:             "clip_caps",
+	CB_TYPE_FILECONTENTS_REQUEST:  "filecontents_request",
+	CB_TYPE_FILECONTENTS_RESPONSE: "filecontents_response",
+	CB_TYPE_LOCK_CLIPDATA:         "lock_clipdata",
+	CB_TYPE_UNLOCK_CLIPDATA:       "unlock_clipdata",
+}
+
+const (
+	// Message flags.
```

Comment:
Could multiple bit be set? cbFlagsMap will atm only map if one bit is set

Reply:
I believe multiple bits could be set, yes.

* here's the spec: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/9e97ce45-0597-43dd-b116-9f62a5b34d54?redirectedfrom=MSDN
* here's the pyrdp parser: https://github.com/GoSecure/pyrdp/blob/63d1b6e0f93a8e1db45b8818c9b0ec9c1374e9cc/pyrdp/parser/rdp/virtual_channel/clipboard.py#L32

It's vague and I understand the original mistake since two out of three messages are OK or FAIL which don't seem compatible.

I'll fix it. I think I have a good idea for an approach.

Reply:
So my initial approach to achieve this was to replicate what you did in 69ec44a. Naively I wrote this:

```go
func parseClipboardData(d *decode.D, length int64) {
	d.FieldStruct("clipboard_data", func(d *decode.D) {
		msgType := uint16(d.FieldU16("msg_type", cbTypesMap))
		d.FieldStruct("msg_flags", func(d *decode.D) {
			d.FieldBool("cb_response_ok")
			d.FieldBool("cb_response_fail")
			d.FieldBool("cb_ascii_names")
		})
		dataLength := d.FieldU32("data_len")
```

This is exactly how the protocol is documented and the `UintMapSymStr` we had before.

![image](https://github.com/wader/fq/assets/546325/5f7a0cf7-c92a-4574-950a-c46e45f57d8c)

![image](https://github.com/wader/fq/assets/546325/38fb39e2-561b-4ce3-b1b2-89c95bc6569d)

Enhancing the protocols' parsers would be straightforward. However, I realized that `FieldStruct` doesn't respect byte order and when I was wondering why, I realized well it can't since it doesn't know the "type" we are about to read...

So to work the parser now looks like this:

```go
func parseClipboardData(d *decode.D, length int64) {
	d.FieldStruct("clipboard_data", func(d *decode.D) {
		msgType := uint16(d.FieldU16("msg_type", cbTypesMap))
		d.FieldStruct("msg_flags", func(d *decode.D) {
			d.FieldRawLen("unused0", 5)
			d.FieldBool("cb_ascii_names")
			d.FieldBool("cb_response_fail")
			d.FieldBool("cb_response_ok")

			d.FieldRawLen("unused1", 8)
		})
		dataLength := d.FieldU32("data_len")
```

In addition to being harder to write, it also has a UX impact with the unnecessary unused fields being exposed to the user:

![image](https://github.com/wader/fq/assets/546325/379a2cc3-f3c0-4870-a6f6-b2466e3a7be1)

This is not how the protocol is described on the wire.

Now, trying to improve this, is there a way we could make FieldStruct or a variant of it byte order aware?

```go
func parseClipboardData(d *decode.D, length int64) {
	d.FieldStruct("clipboard_data", func(d *decode.D) {
		msgType := uint16(d.FieldU16("msg_type", cbTypesMap))
		d.FieldU16Bitfield("msg_flags", func(d *decode.D) {
			d.FieldBool("cb_response_ok")
			d.FieldBool("cb_response_fail")
			d.FieldBool("cb_ascii_names")
		})
		dataLength := d.FieldU32("data_len")
```

Using `FieldU16Bitfield` would under the hood parse the value as a uint16 in little-endian byte order and automatically jump after the 13 unused bits.

If this is too complicated to implement, I think the previous way (const list + `UintMapSymStr`) was better then. Both easier to write and easier to analyze as an analyst (no unnecessary unused fields). However, the bitfield instead of a map problem remains.

Otherwise, is there something else in the decode API I don't see or understand that would be better?

Reply:
> So to work the parser now looks like this:
> 
> ```go
> func parseClipboardData(d *decode.D, length int64) {
> 	d.FieldStruct("clipboard_data", func(d *decode.D) {
> 		msgType := uint16(d.FieldU16("msg_type", cbTypesMap))
> 		d.FieldStruct("msg_flags", func(d *decode.D) {
> 			d.FieldRawLen("unused0", 5)
> 			d.FieldBool("cb_ascii_names")
> 			d.FieldBool("cb_response_fail")
> 			d.FieldBool("cb_response_ok")
> 
> 			d.FieldRawLen("unused1", 8)
> 		})
> 		dataLength := d.FieldU32("data_len")
> ```
> 
> In addition to being harder to write, it also has a UX impact with the unnecessary unused fields being exposed to the user:
> 
> ![image](https://private-user-images.githubusercontent.com/546325/341920247-379a2cc3-f3c0-4870-a6f6-b2466e3a7be1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwNjM1NzAsIm5iZiI6MTcxOTA2MzI3MCwicGF0aCI6Ii81NDYzMjUvMzQxOTIwMjQ3LTM3OWEyY2MzLWYzYzAtNDg3MC1hNmY2LWIyNDY2ZTNhN2JlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMlQxMzM0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZjIyY2U1NzJhZjhmYTNmNjQyZTYzZGFmOTdkMTdiMjY0ZDczMmNlM2I0ODBhOTNkMWJhZWY5YWMwZGUyYTY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.1rhpE9mwTPLglJAz9Y8s-5DofvoBVgiakR169wORiHg)

Yeap this gets a bit messy at times with how fq was designed to work, that is to not hide anything and also give "access" to all bits somehow even unused and unknown stuff.

So in general i've usually opted to let format decoders decode things as detailed as possible to have as few assumption what a end user would like to query and look at, e.g. maybe someone whats to extract or query the unused bits?

> This is not how the protocol is described on the wire.
> 
> Now, trying to improve this, is there a way we could make FieldStruct or a variant of it byte order aware?
> 
> ```go
> func parseClipboardData(d *decode.D, length int64) {
> 	d.FieldStruct("clipboard_data", func(d *decode.D) {
> 		msgType := uint16(d.FieldU16("msg_type", cbTypesMap))
> 		d.FieldU16Bitfield("msg_flags", func(d *decode.D) {
> 			d.FieldBool("cb_response_ok")
> 			d.FieldBool("cb_response_fail")
> 			d.FieldBool("cb_ascii_names")
> 		})
> 		dataLength := d.FieldU32("data_len")
> ```
> 
> Using `FieldU16Bitfield` would under the hood parse the value as a uint16 in little-endian byte order and automatically jump after the 13 unused bits.
> 
> If this is too complicated to implement, I think the previous way (const list + `UintMapSymStr`) was better then. Both easier to write and easier to analyze as an analyst (no unnecessary unused fields). However, the bitfield instead of a map problem remains.
> 
> Otherwise, is there something else in the decode API I don't see or understand that would be better?

All decoding in fq currently is on purpose on bit level made to "hide" byte boundaries, this is to allow decoding bitstream formats, common for media codecs etc, without much fuzz. And as you have noticed this put some burden on some format decoders to byte align and maybe have to deal with padding, unknown or unused bits that a byte-oriented decoder would not care about, fq even automatically adds "gap" fields for bit ranges a decode skips or for some reason can't know about (trailing bits in a format with explicit ending), this is so that no bits are hidden/unreachable. 

So in general if there is a clear bit field it's probably best to decode it they way you did above: a struct with bool/raw fields for each bit even unused or unknown ones. Will be a bit verbose but now all of them are accessible via a jq query. Btw if there bit flag combinations etc that have some special meaning a decode can add "synthetic" fields of any type that will not be "backed" by a bit range but will be visible and queryable (ex samples count in the mp3 decoder https://github.com/wader/fq/blob/master/format/mpeg/mp3_frame.go#L195)

But yeap all this becomes a bit messy and uninvited when decoding little endian bit fields that are > 1 byte :( currently there is no helpers for dealing with it but i have tried to come up with some way of specifying things in the same order as in a spec etc and the let fq figure things out. Actually in the kaitai prototype there is some support for what kaitai calls "bit-endian" that does more or less this. Sadly things get even messier when decoding some bit range inside a little endian integer as now a range a field can "span" two or more bytes and be non-continuous in the bitstream :(. Not sure how to visualise that in the "dump" tree, atm i think i would just opt to let the fiend be the range of the first and the last bit backing the field.

Sorry for wall of text! and hope i managed to explain myself properly and all this is a bit of an unsolved issue with fq so i'm glad someone has input on it.

Reply:
You have made this definitely clearer. Having access to bit-sequence unused at the time the parser was written but that could be used later is a strong argument too. User-friendly before plugin programmer-friendly.